### PR TITLE
Add `InnerProduct` and `CosineExpanded` metric support in NN Descent

### LINKED ISSUE #171: [FEA] Enable inner product distances for NN Descent
IVF-PQ, as a backend of CAGRA, supports inner product metric but at the moment NN Descent does not. It's a fairly straightforward change to update NN Descent to also support inner product metric.

### PR DESCRIPTION
**Pull request by divyegala**
Closes #171

### CODE REVIEW

File: cpp/src/neighbors/detail/nn_descent.cuh [hunk 1]
```diff
@@ -468,26 +472,31 @@ RAFT_KERNEL preprocess_data_kernel(const Data_t* input_data,
...
-    float part_dist = 0;
-    if (idx < dim) {
-      part_dist = s_vec[idx];
-      part_dist = part_dist * part_dist;
-    }
-    __syncwarp();
-    for (int offset = raft::warp_size() >> 1; offset >= 1; offset >>= 1) {
-      part_dist += __shfl_down_sync(raft::warp_full_mask(), part_dist, offset);
+
+  if (metric != cuvs::distance::DistanceType::InnerProduct) {
```
Comment (cjnolet) [inline cpp/src/neighbors/detail/nn_descent.cuh:476]: This seems brittle. We should be able to make this more robust. Doesn't this just apply to L2 metrics?

File: cpp/src/neighbors/detail/nn_descent.cuh [hunk 2]
```diff
@@ -468,26 +472,31 @@ RAFT_KERNEL preprocess_data_kernel(const Data_t* input_data,
...
-    __syncwarp();
   }
 
   for (int step = 0; step < raft::ceildiv(dim, raft::warp_size()); step++) {
     int idx = step * raft::warp_size() + threadIdx.x;
     if (idx < dim) {
-      if (l2_norms == nullptr) {
+      if (metric == cuvs::distance::DistanceType::InnerProduct) {
+        output_data[list_id * dim + idx] = input_data[(size_t)blockIdx.x * dim + idx];
+      } else if (metric == cuvs::distance::DistanceType::CosineExpanded) {
```
Comment (cjnolet) [inline cpp/src/neighbors/detail/nn_descent.cuh:499]: The division and sqrt here seem like they could slow this down quite a bit. Have you done any benchmarks between L2, IP, and Cosine yet to see the differences? I would suggest at least doing a quick comparison, if even only for a couple different parameter settings just to make shis isn't extremely slow.

File: cpp/src/neighbors/detail/nn_descent.cuh (hunk 1 above)
Comment (divyegala) [reply cpp/src/neighbors/detail/nn_descent.cuh:476]: L2 and cosine both as we need the l2 norm. I can change the condition to the exact metrics.

**Comment by divyegala** [comment]
@cjnolet here's some quick benchmark numbers. I printed the metric out to ensure the right one is being passed through from the benchmarks. IMO the impact of the division and `sqrt` is minimal but perhaps you feel differently:

```
metric: cosine
[I] [15:30:05.587420] optimizing graph
[I] [15:30:06.510756] Graph optimized, creating index
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                                        Time             CPU   Iterations        GPU graph_degree index_size intermediate_graph_degree
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
cuvs_cagra.graph_degree32.intermediate_graph_degree64.graph_build_algoNN_DESCENT/process_time/real_time       12.0 s           158 s             1    11.9721           32      1000k                        64 graph_build_algo="NN_DESCENT"
[I] [15:30:06.884944] Overwriting file: /raid/dgala/cuvs/datasets/sift-128-euclidean/index/cuvs_cagra.graph_degree32.intermediate_graph_degree96.graph_build_algoNN_DESCENT
metric: cosine
[I] [15:30:17.943359] optimizing graph
[I] [15:30:19.465069] Graph optimized, creating index
cuvs_cagra.graph_degree32.intermediate_graph_degree96.graph_build_algoNN_DESCENT/process_time/real_time       12.7 s           161 s             1    12.6604           32      1000k                        96 graph_build_algo="NN_DESCENT"
[I] [15:30:19.844759] Overwriting file: /raid/dgala/cuvs/datasets/sift-128-euclidean/index/cuvs_cagra.graph_degree64.intermediate_graph_degree64.graph_build_algoNN_DESCENT
metric: cosine
[I] [15:30:30.708098] optimizing graph
[I] [15:30:32.099818] Graph optimized, creating index
cuvs_cagra.graph_degree64.intermediate_graph_degree64.graph_build_algoNN_DESCENT/process_time/real_time       12.4 s           178 s             1    12.3581           64      1000k                        64 graph_build_algo="NN_DESCENT"
[I] [15:30:32.782952] Overwriting file: /raid/dgala/cuvs/datasets/sift-128-euclidean/index/cuvs_cagra.graph_degree64.intermediate_graph_degree96.graph_build_algoNN_DESCENT
metric: cosine
[I] [15:30:44.454569] optimizing graph
[I] [15:30:46.214032] Graph optimized, creating index
cuvs_cagra.graph_degree64.intermediate_graph_degree96.graph_build_algoNN_DESCENT/process_time/real_time       13.5 s           194 s             1    13.5414           64      1000k                        96 graph_build_algo="NN_DESCENT"
```

```
metric: euclidean
[I] [15:33:53.323826] optimizing graph
[I] [15:33:54.058909] Graph optimized, creating index
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                                        Time             CPU   Iterations        GPU graph_degree index_size intermediate_graph_degree
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
cuvs_cagra.graph_degree32.intermediate_graph_degree64.graph_build_algoNN_DESCENT/process_time/real_time       11.8 s           148 s             1    11.7708           32      1000k                        64 graph_build_algo="NN_DESCENT"
[I] [15:33:54.419808] Overwriting file: /raid/dgala/cuvs/datasets/sift-128-euclidean/index/cuvs_cagra.graph_degree32.intermediate_graph_degree96.graph_build_algoNN_DESCENT
metric: euclidean
[I] [15:34:05.632598] optimizing graph
[I] [15:34:06.974172] Graph optimized, creating index
cuvs_cagra.graph_degree32.intermediate_graph_degree96.graph_build_algoNN_DESCENT/process_time/real_time       12.6 s           160 s             1    12.6361           32      1000k                        96 graph_build_algo="NN_DESCENT"
[I] [15:34:07.347230] Overwriting file: /raid/dgala/cuvs/datasets/sift-128-euclidean/index/cuvs_cagra.graph_degree64.intermediate_graph_degree64.graph_build_algoNN_DESCENT
metric: euclidean
[I] [15:34:18.181052] optimizing graph
[I] [15:34:19.455167] Graph optimized, creating index
cuvs_cagra.graph_degree64.intermediate_graph_degree64.graph_build_algoNN_DESCENT/process_time/real_time       12.2 s           166 s             1    12.2109           64      1000k                        64 graph_build_algo="NN_DESCENT"
[I] [15:34:20.156899] Overwriting file: /raid/dgala/cuvs/datasets/sift-128-euclidean/index/cuvs_cagra.graph_degree64.intermediate_graph_degree96.graph_build_algoNN_DESCENT
metric: euclidean
[I] [15:34:31.725188] optimizing graph
[I] [15:34:33.502260] Graph optimized, creating index
cuvs_cagra.graph_degree64.intermediate_graph_degree96.graph_build_algoNN_DESCENT/process_time/real_time       13.5 s           192 s             1    13.4508           64      1000k                        96 graph_build_algo="NN_DESCENT"
```

**Comment by divyegala** [comment]
gist-960-euclidean numbers:
```
metric: euclidean
[I] [15:50:56.395567] optimizing graph
[I] [15:50:57.138982] Graph optimized, creating index
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                                        Time             CPU   Iterations        GPU graph_degree index_size intermediate_graph_degree
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
cuvs_cagra.graph_degree32.intermediate_graph_degree64.graph_build_algoNN_DESCENT/process_time/real_time       32.5 s           174 s             1    32.4557           32      1000k                        64 graph_build_algo="NN_DESCENT"
metric: euclidean
[I] [15:51:28.327044] optimizing graph
[I] [15:51:29.536358] Graph optimized, creating index
cuvs_cagra.graph_degree32.intermediate_graph_degree96.graph_build_algoNN_DESCENT/process_time/real_time       32.2 s           190 s             1    32.2073           32      1000k                        96 graph_build_algo="NN_DESCENT"
metric: euclidean
[I] [15:52:01.424672] optimizing graph
[I] [15:52:02.593928] Graph optimized, creating index
cuvs_cagra.graph_degree64.intermediate_graph_degree64.graph_build_algoNN_DESCENT/process_time/real_time       33.0 s           188 s             1    32.9537           64      1000k                        64 graph_build_algo="NN_DESCENT"
metric: euclidean
[I] [15:52:34.354768] optimizing graph
[I] [15:52:35.988628] Graph optimized, creating index
cuvs_cagra.graph_degree64.intermediate_graph_degree96.graph_build_algoNN_DESCENT/process_time/real_time       33.0 s           198 s             1    33.0033           64      1000k                        96 graph_build_algo="NN_DESCENT"
```

```
metric: inner_product
[I] [15:56:24.627861] optimizing graph
[I] [15:56:25.362416] Graph optimized, creating index
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                                        Time             CPU   Iterations        GPU graph_degree index_size intermediate_graph_degree
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
cuvs_cagra.graph_degree32.intermediate_graph_degree64.graph_build_algoNN_DESCENT/process_time/real_time       32.4 s           167 s             1    32.4496           32      1000k                        64 graph_build_algo="NN_DESCENT"
[I] [15:56:26.244176] Overwriting file: /raid/dgala/cuvs/datasets/gist-960-euclidean/index/cuvs_cagra.graph_degree32.intermediate_graph_degree96.graph_build_algoNN_DESCENT
metric: inner_product
[I] [15:56:56.738775] optimizing graph
[I] [15:56:57.905511] Graph optimized, creating index
cuvs_cagra.graph_degree32.intermediate_graph_degree96.graph_build_algoNN_DESCENT/process_time/real_time       32.3 s           193 s             1    32.2539           32      1000k                        96 graph_build_algo="NN_DESCENT"
[I] [15:56:58.783456] Overwriting file: /raid/dgala/cuvs/datasets/gist-960-euclidean/index/cuvs_cagra.graph_degree64.intermediate_graph_degree64.graph_build_algoNN_DESCENT
metric: inner_product
[I] [15:57:29.642376] optimizing graph
[I] [15:57:30.942815] Graph optimized, creating index
cuvs_cagra.graph_degree64.intermediate_graph_degree64.graph_build_algoNN_DESCENT/process_time/real_time       32.6 s           198 s             1    32.6481           64      1000k                        64 graph_build_algo="NN_DESCENT"
[I] [15:57:32.024089] Overwriting file: /raid/dgala/cuvs/datasets/gist-960-euclidean/index/cuvs_cagra.graph_degree64.intermediate_graph_degree96.graph_build_algoNN_DESCENT
metric: inner_product
[I] [15:58:02.179546] optimizing graph
[I] [15:58:03.901965] Graph optimized, creating index
cuvs_cagra.graph_degree64.intermediate_graph_degree96.graph_build_algoNN_DESCENT/process_time/real_time       32.5 s           210 s             1    32.4916           64      1000k   
```

```
metric: cosine
[I] [16:01:29.537623] optimizing graph
[I] [16:01:30.389369] Graph optimized, creating index
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                                        Time             CPU   Iterations        GPU graph_degree index_size intermediate_graph_degree
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
cuvs_cagra.graph_degree32.intermediate_graph_degree64.graph_build_algoNN_DESCENT/process_time/real_time       32.5 s           183 s             1    32.4771           32      1000k                        64 graph_build_algo="NN_DESCENT"
[I] [16:01:31.237660] Overwriting file: /raid/dgala/cuvs/datasets/gist-960-euclidean/index/cuvs_cagra.graph_degree32.intermediate_graph_degree96.graph_build_algoNN_DESCENT
metric: cosine
[I] [16:02:01.286174] optimizing graph
[I] [16:02:02.452481] Graph optimized, creating index
cuvs_cagra.graph_degree32.intermediate_graph_degree96.graph_build_algoNN_DESCENT/process_time/real_time       31.7 s           180 s             1    31.7016           32      1000k                        96 graph_build_algo="NN_DESCENT"
[I] [16:02:03.236910] Overwriting file: /raid/dgala/cuvs/datasets/gist-960-euclidean/index/cuvs_cagra.graph_degree64.intermediate_graph_degree64.graph_build_algoNN_DESCENT
metric: cosine
[I] [16:02:34.056109] optimizing graph
[I] [16:02:35.229266] Graph optimized, creating index
cuvs_cagra.graph_degree64.intermediate_graph_degree64.graph_build_algoNN_DESCENT/process_time/real_time       32.5 s           189 s             1    32.4824           64      1000k                        64 graph_build_algo="NN_DESCENT"
[I] [16:02:36.295483] Overwriting file: /raid/dgala/cuvs/datasets/gist-960-euclidean/index/cuvs_cagra.graph_degree64.intermediate_graph_degree96.graph_build_algoNN_DESCENT
metric: cosine
[I] [16:03:06.931169] optimizing graph
[I] [16:03:08.605045] Graph optimized, creating index
cuvs_cagra.graph_degree64.intermediate_graph_degree96.graph_build_algoNN_DESCENT/process_time/real_time       32.8 s           215 s             1    32.8123           64      1000k                        96 graph_build_algo="NN_DESCENT"
```

**Review by cjnolet — APPROVED** [review]

**Pull request merged**